code/datums/components/full_auto_fire.dm

```diff
@@ -0,0 +1,222 @@
+#define AUTOFIRE_TRACKTARGET (1<<0)
+
+
+/datum/component/automatic_fire
+	var/client/clicker
+	var/atom/target
+	var/mouse_parameters
+	var/delay = 0.2 SECONDS //How long between each fire signal.
+	var/delay_timer
+	var/autofire_flags = AUTOFIRE_TRACKTARGET
+	var/shots_fired = 0
+
+
+/datum/component/automatic_fire/Initialize(...)
+	. = ..()
+	if(isgun(parent))
+		RegisterSignal(parent, COMSIG_GUN_FULLAUTOMODE_TOGGLEON, .proc/autofire_turning_on)
+		RegisterSignal(parent, COMSIG_GUN_FULLAUTOMODE_TOGGLEOFF, .proc/autofire_turning_off)
+		return
+	return COMPONENT_INCOMPATIBLE
+
+
+/datum/component/automatic_fire/Destroy()
+	autofire_off()
+	return ..()
+
+
+// There is a gun. Let's configure it so that it turns on whenver the user wields it.
+/datum/component/automatic_fire/proc/autofire_turning_on(datum/source, client/usercli)
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/itemgun_equipped)
+	var/mob/shooter = usercli.mob
+	if(shooter.l_hand == parent || shooter.r_hand == parent)
+		autofire_on(usercli)
+
+
+/datum/component/automatic_fire/proc/autofire_turning_off(datum/source, client/usercli)
+	UnregisterSignal(parent, COMSIG_ITEM_EQUIPPED)
+	autofire_off()
+
+
+// There is a gun and there is a user wielding it. The component now waits for the mouse click.
+/datum/component/automatic_fire/proc/autofire_on(client/usercli)
+	clicker = usercli
+	RegisterSignal(clicker, COMSIG_CLIENT_MOUSEDOWN, .proc/on_mouse_down)
+	RegisterSignal(clicker.mob, COMSIG_MOB_LOGOUT, .proc/autofire_off)
+	RegisterSignal(parent, list(COMSIG_PARENT_QDELETED, COMSIG_ITEM_DROPPED), .proc/autofire_off)
+	parent.RegisterSignal(src, COMSIG_AUTOFIRE_STOPPED, /obj/item/weapon/gun/.proc/on_autofire_stop)

```

Comment:
...what? Just call the proc.

Also you should be passing in this as a callback during component init